[Merged by Bors] - Introduce validator definition file for VC

paulhauner opened the pull request:
## Issue Addressed

NA

## Proposed Changes

- Introduces the `valdiator_definitions.yml` file which serves as an explicit list of validators that should be run by the validator client.
  - Removes `--strict` flag, split into `--strict-lockfiles` and `--disable-auto-discover`  
  - Adds a "Validator Management" page to the book.
- Adds the `common/account_utils` crate which contains some logic that was starting to duplicate across the codebase.

The new docs for this feature are the best description of it (apart from the code, I guess): https://github.com/sigp/lighthouse/blob/9cb87e93ce239bbe06a9d01e92126f978356dd13/book/src/validator-management.md

## API Changes

This change should be transparent for *most* existing users. If the `valdiator_definitions.yml` doesn't exist then it will be automatically generated using a method that will detect all the validators in their `validators_dir`.

Users will have issues if they are:

1. Using `--strict`.
1. Have keystores in their `~/.lighthouse/validators` directory that weren't being detected by the current keystore discovery method.

For users with (1), the VC will refuse to start because the `--strict` flag has been removed. They will be forced to review `--help` and choose an equivalent flag.

For users with (2), this seems fairly unlikely and since we're only in testnets there's no *real* value on the line here. I'm happy to take the risk, it would be a different case for mainnet.

## Additional Info

This PR adds functionality we will need for #1347.

## TODO

- [x] Reconsider flags
- [x] Move doc into a more reasonable chapter.
- [x] Check for compile warnings.

adaszko commented on common/account_utils/src/lib.rs:
```diff
@@ -0,0 +1,129 @@
+}
+
+/// Remove any number of newline or carriage returns from the end of a vector of bytes.
+pub fn strip_off_newlines(mut bytes: Vec<u8>) -> Vec<u8> {
```
[`strip_suffix()`](https://doc.rust-lang.org/std/string/struct.String.html#method.strip_suffix) has been stabilized recently

paulhauner replied:
At the moment nothing stops a user from using *any* bytes that don't end in `/r` or `/n` and I think we'd lose that if we were to convert `bytes` into a `String`.

I think Eth2 clients are in agreement that we should strip newlines but I'm not sure there's consensus on only allowing text. If we start restricting to text-only we might not be able to import keys created by some other client. Because of this, I'm tempted to leave it as is.

adaszko commented on book/src/validator-management.md:
```diff
@@ -0,0 +1,183 @@
+incorrect paths the validator client will still start and attempt to run
+whichever validators it can. As such, care should be taken to observe the logs
+when starting the validator client to ensure it has been able to initialize
+*all* validators.
```
If it isn't too hard to implement: I think the user would expect in this case a hard (and loud) failure if any of the validators hasn't been initialized.

paulhauner replied:
Agreed, I wasn't quite sure about this one. I was originally on your side but I heard another client was still running even if some validators were bad.

I think we might as well stick to our guns here and fail fast and loud :)

paulhauner commented:
I've addressed all your comments @adaszko, thank you!

I also added another two "features" which I've split into individual commits to hopefully make it easy to grok the changes:

- https://github.com/sigp/lighthouse/pull/1357/commits/ce19d629b381d669c856120774afe5cf09bd1864: Allow for a password to be included in the `validator_definitions.yml` file. I feel this might come in useful for some users. I've changed the default permissions for this file to be `600` to reflect that it now (maybe) stores sensitive information.
- https://github.com/sigp/lighthouse/pull/1357/commits/57f430f48ebfcefc64351be476870f06815e4fb2: Allows the `validator_definitions.yml` file to contain either a password path, password or neither. When there is neither we prompt the user for the password. I had some feedback from @carlbeek that he liked this functionality in Nimbus and it also seems like "enter a password each time on boot" is something that might appeal to some people.

I addressed your comments in https://github.com/sigp/lighthouse/pull/1357/commits/6de2fb7f3b94b39d46fdadc88e1fb4cec3c0cc87.

adaszko commented:
Nice 👍

paulhauner commented:
Thanks @adaszko!

bors r+

paulhauner commented:
bors r+
